File: src/main/GO/leetcode/lists/hot100/LeetCode739.go
package hot100

func dailyTemperatures(temperatures []int) []int {
	if len(temperatures) == 0 {
		return nil
	}
	stack := make([]int, 0)
	res := make([]int, len(temperatures))
	stack = append(stack, 0)
	for i := 1; i < len(temperatures); i++ {
		for len(stack) > 0 && temperatures[stack[len(stack)-1]] < temperatures[i] {
			pop := stack[len(stack)-1]
			res[pop] = i - pop
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, i)
	}
	return res
}

File: src/main/java/leetcode/lists/hot100/LeetCode739.java
package leetcode.lists.hot100;

import java.util.Stack;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/19 14:29
 */
public class LeetCode739 {

    public int[] dailyTemperatures(int[] temperatures) {
        if (temperatures == null || temperatures.length == 0) {
            return temperatures;
        }
        Stack<Integer> stack = new Stack<>();
        int[] res = new int[temperatures.length];
        //    1 2 3  6 4  3 2 1 0 5
        stack.push(temperatures[0]);
        for (int i = 1; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int day = stack.pop();
                res[day] = i - day;
            }
            stack.push(i);

        }
        return res;
    }
}
